src/index.ts

```diff
@@ -0,0 +1,241 @@
+
+    if (isNumber(current) || current === '.') {
+      numberAccumulator += current;
+    } else if (current === '-' && i === 0) {
+      numberIsNegative = true;
+    } else if (isOperator(current)) {
+      if (numberAccumulator !== '') {
+        const token = generateNumberToken(numberAccumulator);
+        addNumberToken(token);
+      }
+      if (current === '-' && isOperator(previous)) {
+        numberIsNegative = true;
+      } else {
+        tokenArray.push({
+          type: 'BinaryOperationToken',
+          operation: current,
+        });
+      }
+    } else if (current === '(') {
+      tokenArray.push({
+        type: 'OpeningParenthesis',
+        value: '(',
+      });
+    } else if (isNumber(current) || current === ')') {
+      const token = generateNumberToken(numberAccumulator);
+      addNumberToken(token);
+      tokenArray.push({
+        type: 'ClosingParenthesis',
+        value: ')',
+      });
+    }
+  }
+
+  if (numberAccumulator !== '') {
+    const token = generateNumberToken(numberAccumulator);
+    tokenArray.push(token);
+  }
+
+  return tokenArray;
+};
+
+// This is an implementation of the Shunting-Yard algorithm.
+const transformTokensToReversePolishNotation = (
+  tokenArray: Array<Token>,
+): Array<NumberOrBinaryOperationToken> => {
+  const stack: Array<BinaryOperationToken | OpeningParenthesis> = [];
+  const output: NumberOrBinaryOperationToken[] = [];
+
+  for (let i = 0; i < tokenArray.length; i++) {
+    const currentElement = tokenArray[i];
+
+    if (currentElement.type === 'NumberToken') {
+      output.push(currentElement);
+    } else if (currentElement.type === 'BinaryOperationToken') {
+      if (stack.length === 0) {
+        stack.push(currentElement);
+        continue;
+      }
+
+      const currentPrecedence = operatorPrecedence[currentElement.operation];
+      const previousOperator = stack[stack.length - 1];
+
+      if (previousOperator.type !== 'BinaryOperationToken') {
+        stack.push(currentElement);
+        continue;
+      }
+
+      const previousPrecedence = operatorPrecedence[previousOperator.operation];
+
+      if (currentPrecedence > previousPrecedence) {
+        stack.push(currentElement);
+      } else {
+        while (
+          stack.length &&
+          stack[stack.length - 1].type !== 'OpeningParenthesis'
+        ) {
+          const element = stack.pop();
+          // @ts-expect-error Given the above, we know for sure that this is not an opening parenthesis.
+          output.push(element);
+        }
+        stack.push(currentElement);
+      }
+    } else if (currentElement.type === 'OpeningParenthesis') {
+      stack.push(currentElement);
+    } else {
+      let removedElement = stack.pop();
+
+      if (!removedElement) {
+        throw new Error('Invalid parentheses');
+      }
+
+      while (removedElement.type !== 'OpeningParenthesis') {
+        output.push(removedElement);
+        removedElement = stack.pop();
+      }
+    }
+  }
```

Comment:
Definitely a bit hard to keep track of all the branching logic here. Can we use a `switch` statement at the top level for `currentElement` at the top for a bit more clarity